r-package/grf/R/rank_average_treatment.R

```diff
@@ -0,0 +1,310 @@
+  subset <- validate_subset(forest, subset)
+  # Giving sample weight of 0 is effectively the same as dropping samples.
+  # Do it here with upfront subsetting instead of dealing with it in every bootstrap iteration.
+  subset <- intersect(subset, which(observation.weight != 0))
+  subset.weights <- observation.weight[subset]
+  subset.clusters <- clusters[subset]
+  if (any(forest$W.hat[subset] %in% c(0, 1))) {
+    stop("Cannot compute a doubly robust estimate when some propensities are exactly zero or one.")
+  }
+  if (length(unique(subset.clusters)) <= 1) {
+    stop("The specified subset must contain units from more than one cluster.")
+  }
+  if (!is.null(debiasing.weights)) {
+    if (length(debiasing.weights) == NROW(forest$Y.orig)) {
+      debiasing.weights <- debiasing.weights[subset]
+    } else if (length(debiasing.weights) != length(subset)) {
+      stop("If specified, debiasing.weights must be a vector of length n or the subset length.")
+    }
+  }
+  priorities <- as.data.frame(priorities, fix.empty.names = FALSE)
+  if (ncol(priorities) > 2) {
+    stop("`priorities` should be either a vector or a list/array with two rules.")
+  }
+  empty.names <- colnames(priorities) == ""
+  colnames(priorities)[empty.names] <- c("priority1", "priority2")[1:ncol(priorities)][empty.names]
+  if (anyNA(priorities)) {
+    stop("`priorities` contains missing values.")
+  }
+  if (nrow(priorities) == NROW(forest$Y.orig)) {
+    priorities <- priorities[subset, , drop = FALSE]
+  } else if (nrow(priorities) != length(subset)) {
+    stop("`priorities` must be a vector of length n or the subset length.")
+  }
+  # remap to integers 1, ..., num.unique.prios for quicker tabulate().
+  priorities[,] <- lapply(priorities, function(x) as.integer(as.factor(x)))
+  if (is.unsorted(q, strictly = TRUE) || min(q) <= 0 || max(q) != 1) {
+    stop("`q` should correspond to a grid of fractions on the interval (0, 1].")
+  }
+
+  # For all supported forest types, DR.scores is a subset-length vector
+  # TODO future support for multi_arm_causal_forest could potentially restrict to a single (outcome, contrast).
+  DR.scores <- get_scores(forest, subset = subset, debiasing.weights = debiasing.weights,
+                          compliance.score = compliance.score, num.trees.for.weights = num.trees.for.weights)
+
+  # *** Compute the TOC and RATE ***
+
+  if (target == "AUTOC") {
+    wtd.mean <- function(x, w) sum(x * w) / sum(w)
+  } else if (target == "QINI") {
+    wtd.mean <- function(x, w) sum(cumsum(w) / sum(w) * w * x) / sum(w)
+  }
+
+  # Compute estimates, a function to be passed on to bootstrap routine.
+  # @data: a data.frame with the original data, column 1: DR.scores*sample.weights, column 2: sample.weights,
+  #   column 3: priority scores (integer vector)
+  # @indices: a vector of indices which define the bootstrap sample.
+  # @returns: an estimate of RATE, together with the TOC curve.
+  estimate <- function(data, indices, q) {
```

Comment:
Am also wondering about moving this function definition outside, but keeping it private, if you want "expert" users to be able to call directly into grf:::estimate_rate with their own DR estimates ...